### PR TITLE
Issue 324 - Fix header misalignment regression with "always visible" scrollbar

### LINKED ISSUE #324: Scroll bar mis-alignment bug reappearing in latest Dash-Table version
@chriddyp @Marc-Andre-Rivet @anocella 

Hi,
The bug of scroll bar misalignment (#216 ) is appearing again in the latest Dash Table version. 
This is impacting any upgrade in the Dash-Table for our Apps (ispace). 
Can please provide an update if I am missing something?

Thanks in advance

### PR DESCRIPTION
**Pull request by Marc-Andre-Rivet**
Fix #324.

Virtualization #253 introduced a regression for the table's fragments style generation. Specifically, the style is not generated when the table is not virtualized!

### CODE REVIEW

File: dash_table/package-info.json [hunk 1]
```diff
@@ -39,42 +39,42 @@
...
+    "ts-loader": "^5.3.3",
+    "tslint": "^5.12.0",
     "typescript": "^3.2.2",
     "wait-on": "^3.2.0",
-    "webpack": "^4.27.1",
-    "webpack-cli": "^3.1.2",
-    "webpack-dev-server": "^3.1.10",
+    "webpack": "^4.28.3",
+    "webpack-cli": "^3.2.1",
+    "webpack-dev-server": "^3.1.14",
```
Comment (Marc-Andre-Rivet) [inline dash_table/package-info.json]: This is noise from renovate/all -- we don't commit the build result when updating deps

File: src/dash-table/derived/table/fragmentStyles.ts [hunk 2]
```diff
@@ -16,10 +16,12 @@ export default (
...
     rowPadding: IViewportPadding,
     scrollbarWidth: number
-): { fragment?: CSSProperties, cell?: CSSProperties}[][] => {
+): { fragment?: CSSProperties, cell?: CSSProperties }[][] => {
+    const base = [{}, { fragment: { marginRight: scrollbarWidth } }];
+
     if (!virtualization || !uiCell || !uiViewport) {
         return [
-            [{}, {}],
+            base,
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/derived/table/fragmentStyles.ts]: If there is no virtualization or if the virtualization is not ready -- return the scrolling offset only

File: src/dash-table/derived/table/fragmentStyles.ts [hunk 3]
```diff
@@ -28,12 +30,11 @@ export default (
     const virtualizedHeight = (Math.floor(uiViewport.scrollTop / uiCell.height) - rowPadding.before) * uiCell.height;
     const headersHeight = R.sum(R.map(h => h.height, uiHeaders || []));
 
-    const marginRight = scrollbarWidth;
     const marginTop = virtualization && uiViewport && uiCell ? Math.max(virtualizedHeight - headersHeight, 0) : 0;
     const height = Math.max(fullHeight - marginTop, 0);
 
     return [
-        [{}, { fragment: { marginRight } }],
+        base,
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/derived/table/fragmentStyles.ts]: Otherwise, return offset + virtualization-specific styling

**Comment by bpostlethwaite** [comment]
is it much work to put in a regression test?

File: tests/cypress/tests/unit/derivedFragmentStyles_test.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                true,
+                cell,
+                headers,
+                uiViewport,
+                viewport,
+                padding,
+                scrollbarWidth
+            );
+
+            validate(style, scrollbarWidth);
```
Comment (Marc-Andre-Rivet) [inline tests/cypress/tests/unit/derivedFragmentStyles_test.ts]: Basic unit tests to make sure (not-)virtualized table (not-)ready always returns the correct styling for the header alignment / offset.

**Comment by chriddyp** [comment]
> is it much work to put in a regression test?

or maybe part of the review app for manual testing / qa?

File: cypress.json [hunk 5]
```diff
@@ -5,5 +5,6 @@
     "pluginsFile": "./tests/cypress/plugins/index.js",
     "screenshotsFolder": "./tests/cypress/screenshots",
     "supportFile": "./tests/cypress/support/index.js",
-    "videoFolder": "./tests/cypress/videos"
+    "videoFolder": "./tests/cypress/videos",
+    "numTestsKeptInMemory": 0
```
Comment (Marc-Andre-Rivet) [inline cypress.json]: Some annoying tidbits related to Cypress -- it seems the additional tests tipped it over into Heap Overflow land.

File: dash_table/package-info.json [hunk 6]
```diff
@@ -25,12 +25,15 @@
     "private::wait_js": "wait-on http://localhost:8080",
     "private::opentests": "cypress open",
     "private::runtests-v0": "cypress run --browser chrome --spec 'tests/cypress/tests/server/**/*'",
-    "private::runtests-v1": "cypress run --browser chrome",
+    "private::runtests-v1:unit": "cypress run --browser chrome --spec 'tests/cypress/tests/unit/**/*'",
+    "private::runtests-v1:standalone": "cypress run --browser chrome --spec 'tests/cypress/tests/standalone/**/*'",
+    "private::runtests-v1:server": "cypress run --browser chrome --spec 'tests/cypress/tests/server/**/*'",
+    "private::runtests-v1": "run-s private::runtests-v1:unit private::runtests-v1:standalone private::runtests-v1:server",
```
Comment (Marc-Andre-Rivet) [inline dash_table/package-info.json]: Some annoying tidbits related to Cypress -- it seems the additional tests tipped it over into Heap Overflow land. Break up to test into sections for now. Should probably explore parallelisation in the future.

**Comment by Marc-Andre-Rivet** [comment]
> or maybe part of the review app for manual testing / qa?

Yes! As we plan on linking all core dash-* repos versions together in the future, adding documentation for this in dash-core (https://github.com/plotly/dash-core/blob/master/Manual%20Testing.md) and adding an entry for PR review checklist to that effect (https://github.com/plotly/dash-core/blob/master/PR%20Checklist.md#test-review)

**Pull request merged**